src/basic/json.c

```diff
@@ -0,0 +1,3318 @@
+        return 0;
+}
+
+int json_variant_new_object(JsonVariant **ret, JsonVariant **array, size_t n) {
+        JsonVariant *v;
+        size_t i;
+
+        if (!ret)
+                return -EINVAL;
+        if (n == 0) {
+                *ret = JSON_VARIANT_MAGIC_EMPTY_OBJECT;
+                return 0;
+        }
+        if (!array)
+                return -EINVAL;
+        if (n % 2 != 0)
+                return -EINVAL;
+
+        v = new0(JsonVariant, n + 1);
+        if (!v)
+                return -ENOMEM;
+
+        *v = (JsonVariant) {
+                .n_ref = 1,
+                .type = JSON_VARIANT_OBJECT,
+                .n_elements = n,
+        };
+
+        for (i = 0; i < n; i++) {
+                JsonVariant *w = v + 1 + i;
+
+                w->is_embedded = true;
+                w->parent = v;
+
+                json_variant_set(w, array[i]);
+                json_variant_copy_source(w, array[i]);
+        }
+
+        *ret = v;
+        return 0;
+}
+
+static void json_variant_free_inner(JsonVariant *v) {
+        assert(v);
+
+        if (json_variant_is_magic(v))
+                return;
+
+        json_source_unref(v->source);
+
+        if (v->is_reference) {
+                json_variant_unref(v->reference);
+                return;
+        }
+
+        if (IN_SET(v->type, JSON_VARIANT_ARRAY, JSON_VARIANT_OBJECT)) {
+                size_t i;
+
+                for (i = 0; i < v->n_elements; i++)
+                        json_variant_free_inner(v + 1 + i);
+        }
+}
+
+JsonVariant *json_variant_ref(JsonVariant *v) {
+        if (!v)
+                return NULL;
+        if (json_variant_is_magic(v))
+                return v;
+
+        if (v->is_embedded)
+                json_variant_ref(v->parent); /* ref the compounding variant instead */
+        else {
+                assert(v->n_ref > 0);
+                v->n_ref++;
+        }
+
+        return v;
+}
+
+JsonVariant *json_variant_unref(JsonVariant *v) {
+        if (!v)
+                return NULL;
+        if (json_variant_is_magic(v))
+                return NULL;
+
+        if (v->is_embedded)
+                json_variant_unref(v->parent);
+        else {
+                assert(v->n_ref > 0);
+                v->n_ref--;
+
+                if (v->n_ref == 0) {
+                        json_variant_free_inner(v);
+                        free(v);
+                }
+        }
+
+        return NULL;
+}
+
+void json_variant_unref_many(JsonVariant **array, size_t n) {
+        size_t i;
+
+        assert(array || n == 0);
+
+        for (i = 0; i < n; i++)
+                json_variant_unref(array[i]);
+}
+
+const char *json_variant_string(JsonVariant *v) {
+        if (!v)
+                return NULL;
+        if (v == JSON_VARIANT_MAGIC_EMPTY_STRING)
+                return "";
+        if (json_variant_is_magic(v))
+                goto mismatch;
+        if (v->is_reference)
+                return json_variant_string(v->reference);
```

Comment:
```C
assert_return(v->is_reference, NULL);
```
?

Reply:
nah, a is not a programming error, and assert_return() should only be used for (programming) errors.

a JsonVariant object in this library can either contain data itself, or be a "reference" to another JsonVariant object that contains the actual data. Normally of course, JsonVariants will contain the data directly, however the "reference" logic is used when we put together arrays of variants: if we are supposed to store an array we actually allocate a JsonVariant for each in one single new() call, and then try to store the data directly in each element. But in some cases this cannot work (for example, if the data is an array or object itself, or when it contains a long string that doesn't fit into the JsonVariant space allocated for each element of the array. In that case we'll store a "reference" to another JsonVariant object instead.

This mechanism is kinda neat, as we can efficiently store small variants in arrays but transparently also store longer objects, and this is invisible to the user. 

This logic requires us to follow references first before we use the data in the payload of the JsoNVariant itself.

Reply:
I think I misplaced the comment. I meant to attach it to one of the places which do `goto mismatch` — using assert_return() would avoid the need to do the logging explicitly.